### PR TITLE
COC-8 Add bulk data import and test project

### PR DESCRIPTION
**Pull request by eouw0o83hf**
## Test Project
There's now a test project which validates that the components of the Input process all do what they claim to do. You can run it be `cd`ing into `Take02.Tests` and running `dotnet xunit` from the command line (there's probably something in the VS UI too).

## Data Import
The import process is pretty dumb-simple - there's a `<textarea>` to paste data into and fire it off into the system. The data model is based off of the PoC Google Sheet and tab-delimited, so you can just copy a range of cells and paste them into the `<textarea>`. Be sure not to copy the header row.

### Decisions made
- New `Component`s, `ComponentType`s, `Unit`s, and `Library`s that don't already exist in the database will be added to support the integrity of the rest of the data
- This is a create-only operation: if a `Recipe` already exists in the system, it won't be edited or changed in any way. We can change that in the future, but it keeps the process from screwing anything up.

### Copying data
image

### Getting there
image

### Interface
image

### Result
image

### Proof
Everything from Library down (and components/units as necessary) created by bulk importer
image

/cc @andrewralon

### CODE REVIEW

**Comment by eouw0o83hf** [comment]
**0_o**
image

**Comment by andrewralon** [comment]
+1,365 FTW haha

Was parsing an empty cell value was creating issues? Is that why it's now an "item" when there was no Unit name before? I was wondering how long I could get away with that empty string haha -- NOT LONG.

**Review by andrewralon — APPROVED** [review]
YAY FOR UNIT TESTS TY GG

File: Take02.Tests/Import/RawParserTests.cs [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+
+            var recipeNames = new []
+            {
+                "The GOB",
+                "The Funky Fünke",
+                "Gin-Michael",
+                "Lou's Eel"
+            };
+
+            var ingredientTypes = new []
```
Comment (andrewralon) [inline Take02.Tests/Import/RawParserTests.cs:123]: #classy

File: Take02.Tests/Import/RawParserTests.cs [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+                foreach(var recipe in recipeNames)
+                {
+                    var index = 0;
+                    foreach(var ingredientType in ingredientTypes)
+                    {
+                        importRows.Add(new ImportRow
+                        {
+                            Library = library,
+                            RecipeName = recipe,
+                            Amount = "A lot",
```
Comment (andrewralon) [inline Take02.Tests/Import/RawParserTests.cs:142]: moar code like this. :dancing_men:

File: Take02.Tests/Import/LibraryImporterTests.cs [hunk 3]
```diff
@@ -57,7 +57,7 @@ public LibraryImporterTests()
             });
             await _db.SaveChangesAsync();
 
-            var newLibraryName = "Landie Enterprises";
+            var newLibraryName = "Landie.net: It's dot com";
```
Comment (andrewralon) [inline Take02.Tests/Import/LibraryImporterTests.cs]: bahahaha

File: Take02/ViewModels/ImportViewModel.cs [hunk 4]
```diff
@@ -0,0 +1,11 @@
+using System;
+using System.Collections.Generic;
+
+namespace Take02.ViewModels
+{
+    public class ImportViewModel
+    {
+        public bool AreYouSerious { get; set; }
```
Comment (andrewralon) [inline Take02/ViewModels/ImportViewModel.cs:8]: why so serious

**Review by andrewralon — APPROVED** [review]
Yaaaaaasssss

**Comment by eouw0o83hf** [comment]
@andrewralon 

> Was parsing an empty cell value was creating issues? Is that why it's now an "item" when there was no Unit name before? I was wondering how long I could get away with that empty string haha -- NOT LONG.

Yep - `Ingredient` has a non-nullable FK out to `Unit` and non-nullable `Amount`, so *something* had to go there.

**Pull request merged**